pkg/resourcemanager/controller/health/add.go

```diff
@@ -67,19 +69,63 @@ type ControllerConfig struct {
 // AddToManagerWithOptions adds the controller to a Manager with the given config.
 func AddToManagerWithOptions(mgr manager.Manager, conf ControllerConfig) error {
 	// setup main health reconciler
+	healthReconciler := &reconciler{
+		syncPeriod:   conf.SyncPeriod,
+		classFilter:  &conf.ClassFilter,
+		targetClient: conf.TargetCluster.GetClient(),
+		targetScheme: conf.TargetCluster.GetScheme(),
+	}
+
 	healthController, err := controller.New(ControllerName, mgr, controller.Options{
 		MaxConcurrentReconciles: conf.MaxConcurrentWorkers,
-		Reconciler: &reconciler{
-			syncPeriod:   conf.SyncPeriod,
-			classFilter:  &conf.ClassFilter,
-			targetClient: conf.TargetCluster.GetClient(),
-			targetScheme: conf.TargetCluster.GetScheme(),
-		},
-		RecoverPanic: true,
+		Reconciler:              healthReconciler,
+		RecoverPanic:            true,
 	})
 	if err != nil {
 		return fmt.Errorf("unable to setup health reconciler: %w", err)
 	}
+	healthLogger := healthController.GetLogger()
+
+	if conf.TargetCacheDisabled {
+		// if the target cache is disable, we don't want to start additional informers
+		healthReconciler.EnsureWatchForGVK = func(gvk schema.GroupVersionKind, obj client.Object) error {
+			return nil
+		}
+	} else {
+		watchedObjectGVKs := sync.Map{}
+		healthReconciler.EnsureWatchForGVK = func(gvk schema.GroupVersionKind, obj client.Object) error {
+			// check if we have already added watch for GVK
+			// if not, store GVK in map
+			if _, ok := watchedObjectGVKs.LoadOrStore(gvk, nil); ok {
+				return nil
+			}
+
+			watchedObj := obj.DeepCopyObject().(client.Object)
+			metadataOnly := false
+			if !conf.TargetCluster.GetScheme().Recognizes(gvk) {
+				// If we don't know the GVK, we definitely don't have a special health check for it.
+				// I.e., we only care about whether the object is present or not.
+				// Hence, we can start a metadata-only watch instead of watching the entire object, which saves bandwidth and
+				// memory.
+				metadataOnly = true
+				metadataOnlyObj := &metav1.PartialObjectMetadata{}
+				metadataOnlyObj.SetGroupVersionKind(gvk)
+				watchedObj = metadataOnlyObj
+			}
```

Comment:
What is the purpose of this? Why do we need to watch objects which we don't recognize (and don't have health check functions for)? In below predicate, we even exit early for metadata-only events.

Reply:
Consider this:
Health checks for a `ManagedResource` that contains some unrecognized GVKs (e.g. istio objects) will fail if some of those objects are missing.
I.e., if they are deleted from the system it will result in a failing health check. Health checks would succeed again when they are added back.
However, the health status doesn't change on updates to the objects (we don't have a health check function for it).

That's why we are:
- watching metadata-only for unrecognized GVKs that are part of some `ManagedResources`
- only reacting on create and update events for them